source/common/upstream/load_balancer_impl.cc

```diff
@@ -437,20 +432,32 @@ void RoundRobinLoadBalancer::refresh(uint32_t priority) {
       }
     }
 
+    // Compute schedule offset if we already have been balancing for this
+    // locality.
+    const uint64_t offset =
+        hosts.size() == 0
+            ? 0
+            : (scheduler_.find(source) != scheduler_.end() ? scheduler_[source].rr_index_ : seed_) %
+                  hosts.size();
     // Nuke existing scheduler if it exists.
-    scheduler_[source] = Scheduler{};
-    scheduler_[source].weighted_ = weighted;
+    auto& scheduler = scheduler_[source] = Scheduler{};
+    scheduler.weighted_ = weighted;
     if (weighted) {
       // Populate scheduler with host list.
-      // TODO(htuch): We should add the ability to randomly offset into the host list to
-      // desynchronize the schedule across Envoys in large fleets.
       for (const auto& host : hosts) {
         // We use a fixed weight here. While the weight may change without
         // notification, this will only be stale until this host is next picked,
         // at which point it is reinserted into the EdfScheduler with its new
         // weight in chooseHost().
-        scheduler_[source].edf_.add(host->weight(), host);
+        scheduler.edf_.add(host->weight(), host);
+      }
+      // Cycle through hosts to achieve the intended offset behavior.
+      for (uint32_t i = 0; i < offset; ++i) {
```

Comment:
Would it be sufficient to add the hosts starting at index N and wrapping, rather than adding in-order and doing N picks?

Reply:
Sadly not, since with weighted, the initial insertion order doesn't matter, the weights govern where in the schedule you are.

Reply:
Ok, having thought about this for more than 5 seconds in a meeting :-P  Now I'm not convinced this is particularly fair anyway.  To be truly fair wouldn't we have to loop through between 0-[each host * each weight]?

Check my math because I'm terrible at these.  If you have A w=100, B w=100 C w=1 you'd start with
[A=.01, B=.01, C=1].  
Assuming if our rng told us to max out our picks at 3, we'd pop A, assign it a weight of (1/100+1)/100 -> .0101 so now the quee is
B=.01 A=.0101, C=1 pop B, assign it whatever weight, and you'll still pop C as your third pick.  A should have a 1 in 201 chance of getting picked and I think it deterministically still has a 0% chance of being picked, no?

Reply:
Yes, I came to the same conclusion and have already removed the offset for fairness. I'm still keeping it there to break synchronization and have left a TODO to make it fair.